gammagl/models/specformer.py

```diff
@@ -0,0 +1,298 @@
+
+
+def feature_normalize(x):
+    x = np.array(x)
+    rowsum = x.sum(axis=1, keepdims=True)
+    rowsum = np.clip(rowsum, 1, 1e10)
+    return x / rowsum
+def normalize_graph(g):  # input : g is adj matrix , return the L matrix of graph
+    g = np.array(g)
+    g = g + g.T
+    g[g > 0.] = 1.0
+    deg = g.sum(axis=1).reshape(-1)
+    deg[deg == 0.] = 1.0
+    deg = np.diag(deg ** -0.5)
+    adj = np.dot(np.dot(deg, g), deg)
+    L = np.eye(g.shape[0]) - adj
+    return L  # L = In − D^−1/2 @ A @ D^−1/2
+def eigen_decompositon(g):  #  input : g is adj matrix , return the eigenvalue and eigenvectors of L matrix
+    "The normalized (unit “length”) eigenvectors, "
+    "such that the column v[:,i] is the eigenvector corresponding to the eigenvalue w[i]."
+    g = normalize_graph(g)  #  return L matrix of g
+    e, u = eigh(g)  # return the eigenvalues and eigenvectors of L
+    return e, u
+def transpose_qkv_tlx(X, num_heads):
+    """
+    to split the q,k,v with multiheads
+    :param X: [bsz,query，embed_dim]
+    :param num_heads:
+    :return: [bsz,query，num_heads,embed_dim/num_heads]
+    """
+    X = tlx.ops.reshape(X,
+                        (tlx.ops.get_tensor_shape(X)[0], tlx.ops.get_tensor_shape(X)[1], num_heads, -1))
+    X = tlx.ops.convert_to_tensor(
+        tlx.ops.convert_to_numpy(X).transpose((0, 2, 1, 3))  # make num_heads in front of query
+    )
+    X = tlx.ops.reshape(
+        X, (-1, tlx.ops.get_tensor_shape(X)[2], tlx.ops.get_tensor_shape(X)[3])
+    )
+    return X
+def transepose_output_tlx(X, num_heads):
+    X = tlx.ops.reshape(
+        X, (-1, num_heads, tlx.ops.get_tensor_shape(X)[1], tlx.ops.get_tensor_shape(X)[2])
+    )
+    X = tlx.ops.convert_to_tensor(
+        tlx.ops.convert_to_numpy(X).transpose((0, 2, 1, 3))  # make num_heads in front of query
+    )
+    X = tlx.ops.reshape(
+        X, (tlx.ops.get_tensor_shape(X)[0], tlx.ops.get_tensor_shape(X)[1], -1)
+    )
+
+    return X
+def get_split(y, nclass, seed=0):
+
+    percls_trn = int(round(0.6 * len(y) / nclass))
+    val_lb = int(round(0.2 * len(y)))
+
+    indices = []
+    for i in range(nclass):  # every label
+        h = tlx.convert_to_numpy((y == i))
+        res = np.nonzero(h)
+        res = np.array(res).reshape(-1)
+        index = tlx.convert_to_tensor(res)
+        n = tlx.get_tensor_shape(index)[0]
+        index = tlx.gather( index,np.random.permutation(n) )
+        indices.append(index)
+
+    train_index = tlx.concat(values=[i[:percls_trn] for i in indices], axis=0)
+    rest_index = tlx.concat(values=[i[percls_trn:] for i in indices], axis=0)
+    m = tlx.get_tensor_shape(rest_index)[0]
+    index2 = tlx.convert_to_tensor(np.random.permutation(m))
+    index2 = tlx.cast(index2, dtype=tlx.int64)
+
+    rest_index = tlx.gather(rest_index,index2)
+    valid_index = rest_index[:val_lb]
+    test_index = rest_index[val_lb:]
+
+    return train_index, valid_index, test_index
+# utils
+class DotProductAttention_tlx(tlx.nn.Module):

```

Comment:
I think this class should be put into another file. You may put this into `gammagl/models/attention.py` as other users may use this class. And remember to add the docstring to explain the function and parameters.